test/dist_optimizer_test.py

```diff
@@ -0,0 +1,180 @@
+        self.w = torch.rand((3, 3), requires_grad=True)
+
+    def forward(self, t1):
+        return torch.mm(self.w, t1)
+
+    def get_w(self):
+        return self.w
+
+
+class FailingOptimizer(dist_optimizer.FunctionalOptimizer):
+    def step(self, gradients):
+        raise ValueError('Error running optimizer.')
+
+
+def _call_meth(meth, obj_rref, *args, **kwargs):
+    return meth(obj_rref.local_value().wait(), *args, **kwargs)
+
+
+def remote_meth(meth, obj_rref, *args, **kwargs):
+    """
+    Call rpc.remote on a method in a remote object.
+
+    Args:
+        meth: the method (for example, Class.method)
+        obj_rref (RRef): remote reference to the object
+        args: positional arguments to pass to the method
+        kwargs: keyword arguments to pass to the method
+
+    Returns a RRef to the remote method call result.
+    """
+    return rpc.remote(
+        obj_rref.owner(),
+        _call_meth,
+        args=[meth, obj_rref] + list(args),
+        kwargs=kwargs
+    )
+
+
+def rpc_async_meth(meth, obj_rref, *args, **kwargs):
+    """
+    Call rpc.remote on a method in a remote object.
+
+    Args:
+        meth: the method (for example, Class.method)
+        obj_rref (RRef): remote reference to the object
+        args: positional arguments to pass to the method
+        kwargs: keyword arguments to pass to the method
+
+    Returns a Future to the method call result.
+    """
+    return rpc.rpc_async(
+        obj_rref.owner(),
+        _call_meth,
+        args=[meth, obj_rref] + list(args),
+        kwargs=kwargs
+    )
+
+
+@unittest.skipIf(
+    not torch._six.PY3, "Pytorch distributed autograd package " "does not support python2"
+)
+class DistOptimizerTest(object):
+
+    @property
+    def world_size(self):
+        return 4
+
+    @property
+    def init_method(self):
+        return INIT_METHOD_TEMPLATE.format(
+            file_name=self.file_name, rank=self.rank, world_size=self.world_size
+        )
+
+    @dist_init()
+    def test_dist_optim_exception(self):
+        # distributed version
+        owner1 = 'worker%d' % ((self.rank + 1) % self.world_size)
+        owner2 = 'worker%d' % ((self.rank + 2) % self.world_size)
+
+        remote_module1 = rpc.remote(owner1, MyModule)
+        remote_module2 = rpc.remote(owner2, MyModule)
+        remote_param1 = remote_meth(MyModule.get_w, remote_module1)
+        remote_param2 = remote_meth(MyModule.get_w, remote_module2)
+
+        dst_optim = dist_optimizer.DistributedOptimizer(
+            FailingOptimizer,
+            [remote_param1, remote_param2],
+        )
+
+        with dist_autograd.context() as context_id:
+            torch.manual_seed(0)
+            t1 = torch.rand((3, 3), requires_grad=True)
+            t2 = torch.rand((3, 3), requires_grad=True)
+            output1 = rpc_async_meth(MyModule.forward, remote_module1, t2)
+            output2 = rpc_async_meth(
+                MyModule.forward, remote_module2, output1.wait())
```

Comment:
In followup PRs, we also want to test passing an `RRef` of `output1` to `remote_module2`, right?

Reply:
Correct, I was waiting for the .to_here() autograd propagation PR to be landed. If it's already the case, I can add the test to this PR.

Reply:
It's landed, but we can still add new tests in a PR on top of this. There are some flakiness that we are investigating, which should not block us from landing this PR in its current form.